src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java

```diff
@@ -175,11 +180,26 @@ public StreamingResponseBody streamEvents(
                         nakadiSettings.getDefaultCommitTimeoutSeconds(), client.getClientId());
                 final Subscription subscription = subscriptionDbRepository.getSubscription(subscriptionId);
 
-                authorizationValidator.authorizeSubscriptionRead(eventTypeRepository, client, subscription);
+                final Consumer<String> eventTypeModificationListener = (eventTypeName) -> {
+                    try {
+                        authorizationValidator.authorizeSubscriptionRead(eventTypeRepository, subscription);
+                    } catch (final AccessDeniedException ex) {
+                        final SubscriptionStreamer streamerInstance = streamer.get();
+                        if (null != streamerInstance) {
+                            streamerInstance.stop(ex);
+                        }
+                    }
+                };
+
+                streamer.set(subscriptionStreamerFactory.build(subscription, streamParameters, output,
+                        connectionReady, blacklistService));
+
+                try (Closeable ignore = eventTypeChangeListener.registerListener(
+                        eventTypeModificationListener, subscription.getEventTypes())) {
+                    authorizationValidator.authorizeSubscriptionRead(eventTypeRepository, subscription);
 
-                streamer = subscriptionStreamerFactory.build(subscription, streamParameters, output,
-                        connectionReady, blacklistService, client);
-                streamer.stream();
+                    streamer.get().stream();
+                }
```

Comment:
the same comment as for Low Level controller. Is it somehow possible to separate authorization, because now method `streamEvents` really exploded and it's difficult to read its code.